NumericWidget : Support basic math operators

themissingcow opened the pull request:
This allows a field's existing value to be modified rather than replaced.  Basic value/operator/value expressions are supported using `+`, `-`, `/` and `*`. eg: `1+0.5`, `256 / 2` etc.

andrewkaufman commented:
Should there be a plug tooltip and/or doc change explaining the new feature?

themissingcow commented:
Thanks @andrewkaufman, added `%` for you too, and some docs. 

> Also, targeting 0.57 vs 0.58 ?

As `0.58` is out now, `0.57` is bug-fixes only...

johnhaddon commented on python/GafferUITest/NumericWidgetTest.py:
```diff
@@ -0,0 +1,113 @@
+############################
```
Are we rationing `#` in readiness for a second wave? :)

themissingcow replied:
I was totally out. Fortunately I was able to source some from a neighbouring village. Fixed [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/17324a8a53d899189cf26bd5fac4349a410b79b7)

johnhaddon commented on python/GafferUITest/NumericWidgetTest.py:
```diff
@@ -0,0 +1,113 @@
+
+	def testValueChangedSignal( self ) :
+
+		self.emissions = []
+		def f( w, r ) :
+			self.emissions.append( r )
```
`GafferTest.CapturingSlot` provides a more convenient way of doing this.

themissingcow replied:
Sorry had borrowed this from `TextWidgetTest`, improved [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/17324a8a53d899189cf26bd5fac4349a410b79b7)

johnhaddon commented on python/GafferUITest/NumericWidgetTest.py:
```diff
@@ -0,0 +1,113 @@
+	def testType( self ) :
+
+		w = GafferUI.NumericWidget( 1.0 )
+		w._qtWidget().setText( "2" )
```
`w.setText()`?

themissingcow replied:
Ah yes, good spot, thanks! I think it was too early in the morning when I did this.  Fixed [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/17324a8a53d899189cf26bd5fac4349a410b79b7)

johnhaddon commented on python/GafferUITest/NumericWidgetTest.py:
```diff
@@ -0,0 +1,113 @@
+		self.assertIsInstance( w.getValue(), float )
+
+		w = GafferUI.NumericWidget( 1 )
+		w._qtWidget().setText( "2" )
```
`w.setText()`?

themissingcow replied:
Fixed [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/17324a8a53d899189cf26bd5fac4349a410b79b7)

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -75,7 +76,14 @@ def setValue( self, value ) :
-		return self.__numericType( self.getText() )
+		value = self.getText()
+
+		# Allow derived classes to modify the validator if required
```
I don't understand this comment. What is the expected use case, and if someone does provide a different validator, why do we then ignore it with our `isinstance()` check?

themissingcow replied:
Simplified [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/87cdd4dc907ebdf117de10ef33aec0c3e5421ddf)

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -304,3 +312,70 @@ def __emitValueChanged( self, reason ) :
+
+		return QtGui.QValidator.Acceptable, text, pos
+
+	def eval( self, text ) :
```
As far as I can tell, this is our own method and not part of the QValidator API, correct? Would help to clarify this with a comment I think. It wasn't immediately clear to me that the explicit `eval()` in `getValue()` was the only thing that made this work at all.

themissingcow replied:
Hopefully clarified [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/87cdd4dc907ebdf117de10ef33aec0c3e5421ddf).

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -75,7 +76,14 @@ def setValue( self, value ) :
+		# Allow derived classes to modify the validator if required
+		validator = self._qtWidget().validator()
+		if isinstance( validator, BasicNumericMathExpression ) :
+			value = validator.eval( value )
```
As I understand it, the only reason this new value makes it into the final text is because `__editingFinished()` calls `getValue()` and then does a reformatting. At a minimum, I think we need a comment in `__editingFinished()` to say that this is how expressions get evaluated. But perhaps it might be clearer to do the eval only in `__editingFinished()`? Presumably that would be a problem if someone called `getValue()` directly while there was an expression in there, but we already have an equivalent problem : what should we return when someone calls `getValue()` and there is an incomplete expression, such as `1/`?

themissingcow replied:
I've left it in `getValue` as now we [validate incomplete expressions](https://github.com/GafferHQ/gaffer/pull/3873/commits/4505a343edc81a5073011877110c4b503c8b1efa), we should be in a much better place to always return something sensible. It felt more robust that the expansion would happen when the value was queried without relying on UI events. Happy to change if you'd rather though. Comment added [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/dc49f528883fcf7610c79d042a248453f4027fcb).

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -304,3 +312,70 @@ def __emitValueChanged( self, reason ) :
+
+	def eval( self, text ) :
+
+		if not self.__isExpression( text ) :
+			return text
```
This seems confusing - why sometimes return a string and sometimes a number?

themissingcow replied:
Conformed [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/2b3a1aebbc22fb20c09867ffa5f982595388a119).

johnhaddon commented on python/GafferUI/NumericWidget.py:
```diff
@@ -304,3 +312,70 @@ def __emitValueChanged( self, reason ) :
+
+	def validate( self, text, pos ) :
+
+		if not self.__isExpression( text ) :
+			return self.__typeValidator.validate( text, pos )
+
+		# Invalid chars
+		if re.search( self.__invalid, text ) is not None :
+			return QtGui.QValidator.Invalid, text, pos
+
+		# Remove whitespace to simplify tokenisation
+		text = re.sub( self.__whitespace, "", text )
+
+		# Ends with an operator
+		if re.match( self.__operators, text[ -1 ] ) :
+			return QtGui.QValidator.Intermediate, text, pos
+
+		# Individual numbers invalid for our type validator
+		for number in re.split( self.__operators, text ) :
+			# May be seen with *-1, etc...
+			if number == '' :
+				continue
+			state = self.__typeValidator.validate( number, len(number)-1 )
+			if state[0] != QtGui.QValidator.Acceptable :
+				return state[0], text, pos
+
+		# Some other syntax error
+		try :
+			self.eval( text )
+		except SyntaxError :
+			return QtGui.QValidator.Invalid, text, pos
```
I can't but feel that this could be simpler. We're mixing explicit `__isExpression()` checks, string substitutions, manual splitting , regexes, preexisting validators _and_ calls to `eval()`. Can this just be based around a single `(.+)([+-/*])?(.+)?` regex?

themissingcow replied:
How about [this](https://github.com/GafferHQ/gaffer/pull/3873/commits/4505a343edc81a5073011877110c4b503c8b1efa). It also adds `fixup`, which prevents people leaving the field with a trailing operator. There is however a [bug](https://stackoverflow.com/questions/34055174/qvalidator-fixup-issue#34056852), in the `1.6.0` deps version of the Qt bindings that prevents `fixup` from doing anything (we don't get the `QString` there, so its a lost cause). Fortunately, this is resolved in the `2.0.0` deps, so I've re-based this to `master`.

andrewkaufman replied:
I’m not following the implementation details too closely, but it’d be a pity to defer this to master rather than 0.58, purely to avoid a bug that isn’t todo with Gaffer itself. At IE we’d expect to be able to deploy 0.59 using Qt5.6 anyhow, so for us, the bug won’t even be fixed in 0.59.

themissingcow replied:
In short, you can't stop people leaving trailing operators/invalid expressions in the field without a newer version of `PySide`, so its up to you you and @johnhaddon if you want to have that behaviour in Gaffer or not...

andrewkaufman replied:
I guess a more positive framing is that I’d hope to be able to deploy Gaffer 0.58 with a more modern Qt/PySide and benefit from both the feature and the fix.

That said, it’s obviously not production critical to get this soon, just a nice thing.

themissingcow replied:
I think my main concern is that we'd be leaving in a known issue in stock gaffer, which is used across the entirety of CS and other sites just to support custom builds.

andrewkaufman replied:
From my perspective, we leave in hundreds of known issues in every release of Gaffer. At least this one would already have fix in a future release or a custom build of the same release.

johnhaddon replied:
I've added d7c81c8 which simplifies the logic and adds a workaround for the PySide bug. We'll release this in 0.58.

johnhaddon reviewed (changes requested):
Thanks Tom,
I've made a few comments inline, but also have these notes from a bit of practical testing :

- If I enter `1/` and navigate away, then it remains in that invalid state.
- If I enter `*` and navigate away, then it remains in that invalid state.
- As above, but with any operator and any numeric prefix (I suspect). It looks as if this may be partly a pre-existing problem.
- My naive hope/expectation was that I could double click a field to select all, and enter `*2` to double what was there before. Obviously this doesn't work because I lose the value from the selection as soon as I start typing. I think that's OK as a known limitation if we don't want to overcomplicate things, but in that case I think being allowed to start the string with `*` is misleading.

Perhaps we can deal with this all fairly simply as follows?

- Always require an operator to be preceded by a number (with the exception of `-`, which is meaningful as a unary operator).
- In the event of an incomplete expression being entered, revert to just the number before the operator.
- If `-` is entered, revert to `0`.

> @johnhaddon do you think this kind of feature merits being in the plug tooltip directly or the web docs is fine?

I think it'd be worth adding to the tooltip, below the existing comment about increment/decrement.

Cheers...
John

andrewkaufman commented:
> My naive hope/expectation was that I could double click a field to select all, and enter *2 to double what was there before. Obviously this doesn't work because I lose the value from the selection as soon as I start typing

FWIW, I had the same naive expectation and was a bit disappointed when I realized it didn't work... I wonder how much extra effort it would be to make it so?

themissingcow commented:
> > My naive hope/expectation was that I could double click a field to select all, and enter *2 to double what was there before. Obviously this doesn't work because I lose the value from the selection as soon as I start typing
> 
> FWIW, I had the same naive expectation and was a bit disappointed when I realized it didn't work... I wonder how much extra effort it would be to make it so?

I guess I just expected it to work like it does here as the other apps I've ever had this functionality in work like this.

We could probably do it, but it would be a totally different implementation I think.

themissingcow commented:
> > > My naive hope/expectation was that I could double click a field to select all, and enter *2 to double what was there before. Obviously this doesn't work because I lose the value from the selection as soon as I start typing
> > 
> > 
> > FWIW, I had the same naive expectation and was a bit disappointed when I realized it didn't work... I wonder how much extra effort it would be to make it so?
> 
> I guess I just expected it to work like it does here as the other apps I've ever had this functionality in work like this.
> 
> We could probably do it, but it would be a totally different implementation I think.

Actually no, sorry, brain not working. It would be pretty similar. And we could probably support both. Ill look into it.

themissingcow commented:
> > My naive hope/expectation was that I could double click a field to select all, and enter *2 to double what was there before. Obviously this doesn't work because I lose the value from the selection as soon as I start typing
> 
> FWIW, I had the same naive expectation and was a bit disappointed when I realized it didn't work... I wonder how much extra effort it would be to make it so?

I've not tackled this this time, as it creates another layer of complexity tracking the initial value. We can revisit this if people think it is essential, otherwise trying not to sink tooo much time into this.

johnhaddon commented:
We seem to have a Python 3 test failure : 

```
======================================================================
ERROR: testMaths (GafferUITest.NumericWidgetTest.NumericWidgetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/gaffer/gaffer/build/python/GafferUITest/NumericWidgetTest.py", line 108, in testMaths
    self.assertEqual( w.getValue(), expected )
  File "/__w/gaffer/gaffer/build/python/GafferUI/NumericWidget.py", line 86, in getValue
    return self.__numericType( value )
ValueError: invalid literal for int() with base 10: '0.5'
```

johnhaddon commented:
@themissingcow, I've squashed all your fixups and added my own d7c81c8 which deals with the PySide bug in the way we discussed previously. Could you give it a quick sanity check?

> Should there be a plug tooltip and/or doc change explaining the new feature?

@andrewkaufman, I think so, yes. I've updated d2f94a1 to include it.

themissingcow commented:
> @themissingcow, I've squashed all your fixups and added my own [d7c81c8](https://github.com/GafferHQ/gaffer/commit/d7c81c84c984c3d3887e0043b3645af97d5d7723) which deals with the PySide bug in the way we discussed previously. Could you give it a quick sanity check?

Thanks @johnhaddon, seems to work ok. Do you want to squish your re-work down, the preceding commit has issues on its own.

johnhaddon commented:
Squashed down, and added the `%` operator to the tooltip, although I can't imagine a situation anyone will use it in.

themissingcow commented:
> Squashed down, and added the `%` operator to the tooltip, although I can't imagine a situation anyone will use it in.

LGTM, thanks John, I can't Approve as Its my PR :laughing:

johnhaddon commented:
> We seem to have a Python 3 test failure :

Pushed another version which fixes this prior failure that I assumed had been dealt with already.

themissingcow commented:
> > We seem to have a Python 3 test failure :
> 
> Pushed another version which fixes this prior failure that I assumed had been dealt with already.

I had, but in a fixup which may have been missed in your refactor?

johnhaddon commented:
> I had, but in a fixup which may have been missed in your refactor?

Ugh. I wonder what else I missed?

johnhaddon commented:
> Ugh. I wonder what else I missed?

With the kind assistance of @themissingcow I have confirmed that I haven't lost anything else except some additional tests of my own. I've restored those now, so hopefully this is finally good to go.

themissingcow commented:
The gift that keeps on giving.

`03/2` fails in python 3:

```  File "<string>", line 1
    03//4
     ^
SyntaxError: invalid token
```
@johnhaddon had the bright idea that we should use the exiting validator's fixup for the two numeric parts first, but we need to see if that works in conjunction with the pyside bug.

johnhaddon commented:
> 03/2 fails in python 3: @johnhaddon had the bright idea that we should use the exiting validator's fixup for the two numeric parts first, but we need to see if that works in conjunction with the pyside bug.

Alas, that didn't work, because the validators are happy with a leading `0` anyway. Using the existing validators as we were was really fragile anyway. There was no guarantee they were in agreement with either our hand-rolled expression regex or Python's `eval()`. In fd19b2e I've totally replaced them with a single improved regex, and I've removed the `eval()` too, so there are no further chances for inconsistency. Let's see what I broke this time...

themissingcow reviewed (commented):
Think were there John!

johnhaddon commented:
Thanks Tom, I've squashed the fixups ready for merging.
